src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,227 @@
+            _payloadSerializer.Serialize(writer, message.Item);
+            writer.WriteEndObject();
+        }
+
+        private void WriteInvocationMessage(InvocationMessage message, JsonTextWriter writer)
+        {
+            writer.WriteStartObject();
+            WriteHubMessageCommon(message, writer, InvocationMessageType);
+            writer.WritePropertyName(TargetPropertyName);
+            writer.WriteValue(message.Target);
+
+            if (message.NonBlocking)
+            {
+                writer.WritePropertyName(NonBlockingPropertyName);
+                writer.WriteValue(message.NonBlocking);
+            }
+
+            writer.WritePropertyName(ArgumentsPropertyName);
+            writer.WriteStartArray();
+            foreach (var argument in message.Arguments)
+            {
+                _payloadSerializer.Serialize(writer, argument);
+            }
+            writer.WriteEndArray();
+
+            writer.WriteEndObject();
+        }
+
+        private static void WriteHubMessageCommon(HubMessage message, JsonTextWriter writer, int type)
+        {
+            writer.WritePropertyName(InvocationIdPropertyName);
+            writer.WriteValue(message.InvocationId);
+            writer.WritePropertyName(TypePropertyName);
+            writer.WriteValue(type);
+        }
+
+        private InvocationMessage BindInvocationMessage(JObject json, IInvocationBinder binder)
+        {
+            var invocationId = json.Value<string>(InvocationIdPropertyName);
+            var target = json.Value<string>(TargetPropertyName);
+
+            var nonBlocking = false;
+            var nonBlockingProp = json.Property(NonBlockingPropertyName);
+            if (nonBlockingProp != null)
+            {
+                nonBlocking = nonBlockingProp.Value.ToObject<bool>();
+            }
+
+            var paramTypes = binder.GetParameterTypes(target);
+            var arguments = new object[paramTypes.Length];
+
+            var args = json.Value<JArray>(ArgumentsPropertyName);
+            for (var i = 0; i < paramTypes.Length; i++)
+            {
+                var paramType = paramTypes[i];
+
+                // TODO(anurse): We can add some DI magic here to allow users to provide their own serialization
+                // Related Bug: https://github.com/aspnet/SignalR/issues/261
+                arguments[i] = args[i].ToObject(paramType, _payloadSerializer);
+            }
+
+            return new InvocationMessage(invocationId, target, arguments, nonBlocking);
+        }
+
+        private StreamItemMessage BindResultMessage(JObject json, IInvocationBinder binder)
+        {
+            var invocationId = json.Value<string>(InvocationIdPropertyName);
+            var payload = json.Value<JToken>(ResultPropertyName);
+
+            var returnType = binder.GetReturnType(invocationId);
+            return new StreamItemMessage(invocationId, payload?.ToObject(returnType, _payloadSerializer));
```

Comment:
I actually don't think I like requiring the property to be present even when the value is null. Certainly our clients/servers can work that way, but most JSON formats generally consider a missing property to be equal to `null`, unless there's a specific reason to treat them differently.